design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. To avoid this, we will parse over each K8s Secret and perform a batch retrieval request. That way, if there is a failure in retrieving a Conjur secret (permissioning for example), it will be contained to that secret and not impact all the secrets in the namespace.
+
+```pseudocode
+for each k8s-secret
+  variable-ids = k8s-secret.parseConjurMap() # list of paths of key in ConjurMap
+  GET /secrets{?variable-ids}
+  # Log failure or success
+```
+
+Variable IDs will resemble the following:
+
+```
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-credentials
+type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets/db_username # the value is one element in the variable-ids list
+    password: secrets/db_password # the value is one element in the variable-ids list
+```
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Elaboration                                                  | Pros                                                         | Cons                                        |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- |
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Scalable <br /><br />- Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get K8s secrets name by configuration in Secrets Provider    | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we chose the first solution, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
```

Comment:
please address backwards compatibility on this issue. what happens if a user sets the required k8s secrets with the existing environment variable and not with the label?

Reply:
also - i am not sure that the UX has improved in this case. If i have 1000 k8s secrets that i want to update and one secrets-provider then instead of setting one environment variable i now need to modify 1000 manifests of secrets.